### DISCUSSION TITLE
Building ecosystems and pipelines around S3 classes

### DISCUSSION THREAD

**Discussion by joshwlambert** [epiverse-trace/epiverse-trace.github.io#224] [Blog post ideas]
I propose writing a blog post about the benefit of designing S3 classes in R that can be used as a unit around which interoperable ecosystems (R packages in this case) can be built. 

The post will have a small introduction to S3 classes in R, and then give a couple of examples of instances where S3 classes have been used as a central point of reference for methods development. My first thoughts on which packages and classes to discuss are:

- `phylo` class from the {ape} R package
- `incidence` class from the {incidence2} R package
- `gg` class from {ggplot2} R package (this one I'm unsure of the details so would have to do some research to be sure of the specifics and make sure that the `gg[package]` derivative packages are also using the class).

The post would end with some details around {epiparameter} and the proposal for the `epidist` class to become a central unit which epidemiological methods packages can adopt.

I think this would follow on nicely from @Bisaloo's recent [post on S3 generics](https://epiverse-trace.github.io/posts/s3-generic/). 

Please leave a comment with thoughts on the viability and usefulness of the potential post and whether you know of other classes that would be useful to include.

**Comment by Bisaloo** [comment]
Initial thoughts:

- should this focus on S3, or stay generic and treat also the case when S4 classes, which are common pieces of infrastructure in spatial analysis (e.g., raster / terra)? If you focus on S3, you will probably be expected to explain this choice which will probably distract from the main message of this post.
- sf is another good example for S3
- should we try to make this post in collaboration with Emmanuel Paradis (not sure how keen he would be)? The main question here IMO is: "how are standards born?" and I'm not completely sure we can answer this question without internal insight. Although Emmanuel might not know either :shrug:

**Comment by joshwlambert** [comment]
Thanks for the input @Bisaloo. 

> should this focus on S3, or stay generic and treat also the case when S4 classes, which are common pieces of infrastructure in spatial analysis

I would be happy to include examples of where S4 is used. Especially across bioconductor and bioinformatics where S4 is more prevalent. However, the problem then extends to having to explain why S3 & and S4 are included, but R6, R7 and RC are not included. I don't think it's an issue and would just require a single sentence explaining the scope of the post. 

> sf is another good example for S3

Yes, that would be good to include. @pratikunterwegs has lots of spatial expertise so I can consult him once I've got a draft. I think we should not include too many classes, 2-3 should adequately explain the point we are trying to make. Let me know if you have a priority list of classes.

> should we try to make this post in collaboration with Emmanuel Paradis

Sure, I'd be happy to reach out and ask if he would like to collaborate. I'll send him an email and inquire.

**Comment by pratikunterwegs** [comment]
Happy to chip in as required. Agree that classes from {terra} and {sf} packages are good candidates, although their older analogues {raster} and {sp} might be more useful (S4) case studies as they have been around for longer. A good example is an [MPI-AB initiative called MoveApps](https://www.moveapps.org/), that seeks to build data processing pipelines around the {move} package which is built on top of {sp} if I recall correctly.
